src/AutoRest.CSharp/Mgmt/Generation/MgmtClientBaseWriter.cs

```diff
@@ -107,44 +112,101 @@ protected void WriteDebugValidate(CodeWriter writer)
         {
-            writer.Line($"private readonly {typeof(ClientDiagnostics)} {ClientDiagnosticsField};");
+            string readOnly = useReadonly ? " readonly" : string.Empty;
+            if (!useMultipleClientDiagnostics)
+                writer.Line($"private{readOnly} {typeof(ClientDiagnostics)} {ClientDiagnosticsField};");
             foreach (var client in clients)
             {
                 // we might have multiple rest client field, if we combined multiple request together in one resource in case of extension resources
-                writer.Line($"{RestClientAccessibility} readonly {client.Type} {GetRestClientVariableName(client)};");
+                if (useMultipleClientDiagnostics)
+                    writer.Line($"private{readOnly} {typeof(ClientDiagnostics)} {GetClientDiagnosticFieldName(client)};");
+                writer.Line($"{RestClientAccessibility}{readOnly} {client.Type} {GetRestFieldName(client)};");
             }
         }
 
         protected void WriteRestClientConstructionForResource(
             Resource resource,
             IEnumerable<MgmtRestClient> restClients,
-            Func<MgmtRestClient, string> getSubId,
+            string subIdVaraiable,
             string clientDiagVariable,
-            string optionVariable,
+            string diagnosticOptionsVariable,
+            string armClientVariable,
             string pipelineVariable,
             string uriVariable,
             string accessType,
             bool writeVariable)
         {
             var nameSpace = $"{resource.Type.Namespace}.Models";
             _writer.UseNamespace(nameSpace);
-            _writer.Line($"{optionVariable}.TryGetApiVersion({resource.Type.Name}.ResourceType, out string apiVersion);");
+            _writer.Line($"{armClientVariable}.TryGetApiVersion({resource.Type.Name}.ResourceType, out string apiVersion);");
             foreach (var restClient in restClients)
             {
                 var variableDeclaration = writeVariable ? $"{restClient.Type.Name} " : string.Empty;
-                _writer.Line($"{variableDeclaration}{GetRestClientVariableName(restClient)} = {accessType}{restClient.Type.Name}({clientDiagVariable}, {pipelineVariable}, {optionVariable}{getSubId(restClient)}, {uriVariable}, apiVersion);");
+                _writer.Line($"{variableDeclaration}{GetRestFieldName(restClient)} = {GetRestConstructorString(accessType, restClient, clientDiagVariable, pipelineVariable, diagnosticOptionsVariable, subIdVaraiable, uriVariable, ", apiVersion")};");
             }
         }
 
-        protected virtual string GetRestClientVariableName(RestClient client)
+        protected string GetProviderNamespaceFromReturnType(string? returnType)
+        {
+            if (returnType is null)
+                return "_defaultRpNamespace";
+
+            var data = Context.Library.ResourceData.FirstOrDefault(data => data.Declaration.Name == returnType);
```

Comment:
Why is it comparing `Name` instead of `Type`?

Reply:
because we only care about the name here, the type equals overload checks for a few other things